### PR TITLE
Add more combinators to Data.Optional

### PR DESCRIPTION
**Pull request by aiya000**
Hi :smile:  
I want to add .first(), .last(), .flat_map(), and .optional() to Data.Optional.

How would you think about this? :thinking:

### CODE REVIEW

**Comment by ujihisa** [comment]
Hmm, it's just for list inside Data.Optional right? It believe it shouldn't have these short `first/last` names but should have long descriptive names such as `fmap_first` or `fmap_last`, so that users can tell that they are lifted functions.

File: autoload/vital/__vital__/Data/Optional.vim [hunk 1]
```diff
@@ -127,6 +127,19 @@ function! s:flatten(o, ...) abort
...
+    for x in a:xs
+        if s:exists(x)
+            return x
+        endif
+    endfor
+    return s:none()
+endfunction
+
+function! s:last(xs) abort
+    return s:first(reverse(a:xs))
```
Comment (thinca) [inline autoload/vital/__vital__/Data/Optional.vim]: `reverse()` changes the argument directly, so this has side effect.

**Comment by aiya000** [comment]
@thinca Thanks!
I probably fixed :dog2:

**Comment by aiya000** [comment]
TODOっぽいメモ:
```vim
let xx = なんらかのoptional
echo O.optional(xx,
    \ { x -> Msg.echo('Yes! ' . x) },
    \ { x -> Msg.echo('No... :(') }
\ )
" xxがnoneなら 'No... :(` が
" xxがsome(10)なら 'Yes! 10' が 表示される
```
これ欲しい

**Comment by aiya000** [comment]
Sorry, I added `.flat_map()` and `.optional()` at now,
and updated this PR's description :bow:

**Comment by ujihisa** [comment]
I `git rebase master`ed

**Comment by ujihisa** [comment]
@aiya000 I'm sorry but can you resolve conflicts?

**Comment by aiya000** [comment]
@ujihisa of course!
I'll try later :smile:

**Comment by aiya000** [comment]
`rebase`d!

**Comment by ujihisa** [comment]
:clap:

**Pull request merged**